aten/src/ATen/native/TensorShape.cpp

```diff
@@ -193,9 +192,8 @@ Tensor & _cat_out_cpu(TensorList tensors, int64_t dim, Tensor& result) {
   result_size[dim] = cat_dim_size;
 
   // skip resizing if size of result is same as expected
-  if (result.sizes() != result_size) {
-    result.resize_(result_size, first_tensor_mem_format);
-  }
+  // raise a warning while resizing if output has one or more elements
+  at::native::resize_output(result, result_size, first_tensor_mem_format);
```

Comment:
Using resize_output is correct here, to preserve the current behavior of the memory format of the output being derived from the first tensor when a user doesn't specify `out=` then I think the other callers of this function, like here:

https://github.com/pytorch/pytorch/blob/3a88e3fa3176b39dcb5aff9c995b3107a14820ff/aten/src/ATen/native/TensorShape.cpp#L268

would need to change. In particular, for this issue they could allocate an output of the correct size and with the desired memory format in advance. Then resize_output wouldn't resize them. 

This is similar to what structured kernels do, too. cc @bdhirsh to see if we could make cat a structured kernel or not to resolve this issue.

Reply:
@ngimel and I had an opportunity to review further offline. While dtype is actually computed in advance, it's true that precomputing the shape would be redundant. We also realized we need to further clarify our handling of out= args in the developer wiki.

So here's what we recommend:

- use resize_output_check directly
- if the output needs to be resized, resize it with the memory format of the first tensor (like today)
- if the output doesn't need to be resized, preserve its memory format (don't do anything to it)

This will need a comment to explain why we're not calling resize_output directly

Reply:
Thanks, @mruberry @ngimel for reviewing this extensively, especially the resize behaviour in cat_out. I'll address other specific comments in a bit (need a second read to make sure I understand everything correctly). 

IIUC (and I might be jumping the gun here), the final decision looks like the behaviour I had in this [commit](https://github.com/pytorch/pytorch/pull/62560/commits/7df250c421aa0d1e916e316c36d09d6c526c1fcd) previously, but instead of actually writing the warning in here (see how I did it previously in that commit), it's better to use the `resize_output_check` function declared in`Resize.h` which will warn if required.

Reply:
I believe your understanding is correct; sorry for the confusion and thanks for clarifying

Reply:
Awesome! Absolutely no worries, always good to see these discussions (learnt a few new things and also motivates me to look at how the porting to structured kernels is done, maybe sometime later).

 Will go ahead with this approach then. :))